HW-1 [1주차과제] Naver API 연동 앱 구현

gitpushyj opened the pull request:
Naver API 연동 앱 구현

gitpushyj committed:
HW-1 의존 라이브러리 추가

gitpushyj committed:
HW-1 인터넷 접근 권한 추가

gitpushyj committed:
HW-1 메인화면 구성

gitpushyj committed:
HW-1 응답 모델 정의

gitpushyj committed:
HW-1 Retrofit 설정

gitpushyj committed:
HW-1 Naver api 요청 인터페이스 정의

gitpushyj committed:
HW-1 검색 레이아웃 및 기능 구현

gitpushyj committed:
HW-1 스와이프 애니메이션 제거

gitpushyj committed:
HW-1 웹뷰 연동

gitpushyj committed:
HW-1 영화 검색 카드 디자인 정보 변경

gitpushyj committed:
HW-1 에러시 Toast 랜더링

gitpushyj committed:
HW-1 바텀 아이콘 변경

gitpushyj committed:
HW-1 string 리소스화

gitpushyj committed:
HW-1 ListAdapter 추상화

gitpushyj committed:
HW-1 패키지 수정

gitpushyj committed:
HW-1 MovieFragment 리팩토링

gitpushyj committed:
HW-1 리팩토링

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/NaverApiProvider.kt:
```diff
@@ -0,0 +1,37 @@
+val naverApi: NaverRequestApi
+    get() = Retrofit.Builder()
+        .baseUrl("https://openapi.naver.com/v1/search/")
+        .addCallAdapterFactory(CoroutineCallAdapterFactory())
```
baseUrl에 api version이 포함되는 건 가급적이면 피하는게 좋습니다. 
실제 같은 요청 성격에 API에 versioning이 되면 클라이언트 앱에서 구분이 필요합니다.
version을 포함하여 이후 주소는 REST method에 선언하는게 좋은 구조로 생각됩니다.

gitpushyj replied:
그것까진 생각 못했었네요 ㅎㅎ 
감사합니다 반영했어요!
cbf4e0b47c65cfb43400b6c37eb973a796430bfd

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/api/NaverRequestApi.kt:
```diff
@@ -0,0 +1,39 @@
+
+interface NaverRequestApi {
+
+    @GET("movie.json")
```
endpoint에는 version을 포함하는게 좋을 것 같습니다.

gitpushyj replied:
바로 위에 주신 피드백과 동일한 맥락이라 같은 커밋으로 남겼습니다~!

cbf4e0b47c65cfb43400b6c37eb973a796430bfd

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,50 @@
+    abstract fun setListViewAdapter(listView: RecyclerView)
+
+    protected open fun fetchSearchedList() {
+        val adapter = listView.adapter
```
val adapter = listView.adapter ?: return

adpater가 null 인 경우에는 아래의 코드 진행이 무의미 해보입니다. 위와 같이 처리하는건 어떨까요?

gitpushyj replied:
그러네요. 어차피 무의미 한 코드를 실행하느니 return 시키는게 효율적이겠네요!
감사합니다 ㅎㅎ

7869674cc300f86cec2e3a5f77ec1ed4d921018c

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/image/ImageListAdatper.kt:
```diff
@@ -0,0 +1,68 @@
+                fragmentActivity.startWebView(image.link)
+            })
+        }
+        if (position > itemCount - PAGING_OFFSET) loadMore(itemCount)
```
ViewHolder를 구분하는게 좋지 않을까 생각이 듭니다. 한마디로 말하자면 역할 분담(?)과 같은건데....
getItemViewType에서 인덱스를 체크 해서 마지막 페이지 일 때는 (별도 선언) LoadingViewHolder를 생성하여 로드 하는 방식으로 변경하면 onBindViewHolder에서 코드가 간결할 것 같습니다.

```
override fun onBindViewHolder(holder: ViewHolder, position: Int) =
            when (holder) {
                is ImageViewHolder -> {
                    //  bind
                }
                is LoadingViewHolder -> {
                    // loadmore
                }
                else -> {
                    throw Exception("Unknown holder: ${holder::class}")
                }
            }
```

gitpushyj replied:
확실히 가독성이 좋아지네요!! ㅎㅎ
사실 저도 고민해본 부분이긴 한데요... 결국 추가 로딩이 필요한 position에 도달했을 때, 데이터를 추가 로딩 할 뿐만 아니라 그 position에 해당하는 데이터도 뿌려줘야 하기 때문에  LoadingViewHolder가 데이터를 뿌려주는 일반 ViewHolder를 중복 생성하는 꼴이 되버릴 것 같더라구요.

onBindViewHolder가 길어지기도 하고, 동일한 뷰를 그리는 두 개의 ViewHolder를 만드는게 조금 걸려서 보류중이었는데 그래도 LodingViewHolder를 구현 하는게 더 나으려나요..?ㅎㅎ

JayYi replied:
사실 현재 작성하신 코드에 큰 문제는 없어 꼭 수정이 필요하다고는 생각이 들지 않아요.
단지, 더 좋은 코드를 작성하실 수 있을 것 같아 말씀드린거였습니다. 
개선이 필요하다고 느끼시면 나중에 시간 나실 때 개선해보시고 리뷰가 필요하다면 저한테 말씀해주세요.
그리고, 만약 제가 개선을 하게 된다면 아래와 같이 작성하지 않을 까 생각이 드네요.
```
  fun getItemViewType = list size < 0 ImageViewHolder, 아닐 경우 LoadingViewHolder
  fun onBindViewHolder = when (holder)....위의 답변 과 같이 작성
  fun getItemCount = list empty 이면 0, 아닐 경우 list size +1

 위와 같이 작성하면 LoadingViewHolder에 데이터를 받아 뷰를 그릴 필요가 없어짐
  LoadingViewHolder에 requested와 같은 변수(default=false)를 두고 변수 값에 따라 서버에 데이터 요청을 하면 될 것 같습니다.
```
우선은 answer needed  라벨을 해놓을께요. 고민해보시고 추후 수정하시겠다고 하시면 답변 남겨 주시면  제가 approved 하고 라벨 변경 하겠습니다.

gitpushyj replied:
음 getItemCount에서 +1을 해줌으로써 LoadingViewHolder에 데이터를 뿌리지 않는게 가능한건가요...
말씀대로 하면 onBindViewHolder는 가독성이 높아지지만  viewType 구분이 필요한 메서드들의 코드량 증가량과 이해해야 하는 로직이 많아질것 같아서요ㅠ 물론 그려야 하는 리스트의 각 아이템 UI가 지금처럼 한가지가 아닌 그 이상일 경우에는 어차피 구분을 해야하니 어쩔수 없지만...
이 부분은 제가 이해를 잘 못해서 생기는 것일 수도 있으니, 토요일에 직접 들어보고싶습니다..ㅎㅎ

JayYi replied:
알겠습니다. 라벨 변경해주시면 approved해서 석준님 리뷰어로 추가 할께요.

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchListAdapter.kt:
```diff
@@ -0,0 +1,50 @@
+        fragmentActivity.lifecycleScope.launch {
+            try {
+                val addedList = getMoreItemListFromStartIndexAsync(textOnEditTextView, itemCount + 1)
+                currentList.toMutableList().let { list ->
```
?가 빠진것 같습니다...
추가로 제가 코루틴을 잘 몰라서...currentList 는 어디에서 호출된건가요.......

gitpushyj replied:
currentList는 부모 클래스인 ListAdapter에서 제공되는건데 Nullable하지 않은 값이라서 ?처리가 필요없습니다.
다만 let이 주로 ?.let 형태로 많이 쓰이다 보니 어색해 보일수도 있겠네요..ㅎㅎ
run으로 변경했습니다!

426bdef475d96051497b6cae3a1ded8ac69c07bb

JayYi commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/webview/WebviewActivity.kt:
```diff
@@ -0,0 +1,41 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_webview)
+
+        webViewUrl = intent?.extras?.getString(URL) ?: ""
```
intentp url에 포함되지 않으면 WebView를 load하지 않고 별도의 예외 처리 하는게 좋을 것 같다는 생각이듭니다.
(의미 없는 빈스트링의 url를 로딩하는게 무의미해보여요 ㅠㅠ)

gitpushyj replied:
반영했습니다. 감사합니다~!

e565bf2a4ee4434f086dd2d8969e907916833996

JayYi commented on wooooooak/StudyApp/app/src/main/res/layout/activity_main.xml:
```diff
@@ -4,15 +4,25 @@
+        android:layout_height="0dp"
+        app:layout_constraintBottom_toTopOf="@+id/bottom_nav"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
```
저도 그런것 같긴한데.....
width가 match라 제외하셔도 될 것 같습니다.

gitpushyj replied:
오.... 그렇네요 생각해보니 match면 제약을 걸어주지 않아도 되겠네요 ㅎㅎㅎ
감사합니다~!^^

f270e16a1f7b130fdf2d6bb8d0156afa54685bce

JayYi reviewed (changes requested):
리뷰완료~! 확인 부탁드려요

gitpushyj committed:
HW-1 baseUrl의 version 정보 리팩토링

gitpushyj committed:
HW-1 BaseFragment의 adapter가 null일 경우 처리

gitpushyj committed:
HW-1 BaseSearchListAdapter 리팩토링

gitpushyj committed:
HW-1 웹뷰의 url이 null일 경우 처리

gitpushyj committed:
HW-1 activity_main 리팩토링

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/model/response/blog/BlogList.kt:
```diff
@@ -0,0 +1,27 @@
+
+import com.google.gson.annotations.SerializedName
+
+data class BlogList(
```
4개 모델이 각각 유사한 형태로 되어 있습니다.
```
data class NaverQueryResponse<T>(
    @SerializedName("lastBuildDate")
    val lastBuildDate: String,
    @SerializedName("total")
    val total: Int,
    @SerializedName("display")
    val display: Int,
    @SerializedName("items")
    val items: List<T>
)
```
처럼 만들어서 사용하면 될 것 같습니다.

gitpushyj replied:
감사합니다. 반영했습니다!
04299c586131fd545c10220c8e22ace70b20a6a9

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,48 @@
+
+abstract class BaseSearchFragment : Fragment() {
+
+    abstract val layoutId: Int
```
생성자를 통해서 전달 받으면 됩니다.

gitpushyj replied:
수정했습니다!
741d493d0316f00e1667e71ad407268dd7e40a40

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,48 @@
+abstract class BaseSearchFragment : Fragment() {
+
+    abstract val layoutId: Int
+    private lateinit var fragmentView: View
```
fragmentView를 가지고 있을 필요가 없습니다.
onCreateView() 이후 시점에서 view를 통해서 접근할 수 있습니다

gitpushyj replied:
아 그러네요 ㅎㅎ
감사합니다 ~!

6acc2f3671995388db59265eb947b908a41ecda2

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,48 @@
+        savedInstanceState: Bundle?
+    ): View? {
+        fragmentView = inflater.inflate(layoutId, container, false)
+        listView = fragmentView.list_view
```
onCreateView에서는 정말 view를 생성하는 작업만 하면 됩니다.

나머지 작업은 onActivityCreated안에서해주세요.

gitpushyj replied:
넵, 반영했습니다.

abb655d8d818e073f3299cc42907341070a6ff8d

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,48 @@
+
+    abstract val layoutId: Int
+    private lateinit var fragmentView: View
+    private lateinit var listView: RecyclerView
```
listView대신 recyclerView라고 해주세요

gitpushyj replied:
반영했습니다 ㅎㅎ
c3e42439fc222db0b7c85e5c36b4d8be40b82ff3

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/base/BaseSearchFragment.kt:
```diff
@@ -0,0 +1,48 @@
+
+    abstract fun setListViewAdapter(listView: RecyclerView)
+
+    protected open fun fetchSearchedList() {
```
```kotlin
protected open fun fetchSearchedList() {
    val inputText = inputTextView.text.toString()
    if (inputText.isNotEmpty()) {
        (listView.adapter as? Searchable)?.searchByTitle(inputText)
    }
}
```
조금 더 간결하게 사용할 수 있습니다

gitpushyj replied:
아 이런방법이 있었네요 ㅎㅎ 감사합니다.
471fc47c3f14dfc34c7a51d321b7deed64836423

sjjeong commented on wooooooak/StudyApp/app/build.gradle:
```diff
@@ -20,15 +20,39 @@ android {
+    implementation "com.squareup.retrofit2:retrofit:2.6.0"
+    implementation "com.google.code.gson:gson:2.8.5"
+    implementation "com.squareup.retrofit2:converter-gson:2.6.0"
+    implementation 'com.jakewharton.retrofit:retrofit2-kotlin-coroutines-adapter:0.9.2'
```
https://github.com/JakeWharton/retrofit2-kotlin-coroutines-adapter
retrofit:2.6.0를 사용한다면 해당 dependency는 불필요합니다.

gitpushyj replied:
수정했습니다~!

c0fedef4c9a636b19307adb964925ae3046315f9

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/blog/BlogListAdatper.kt:
```diff
@@ -0,0 +1,69 @@
+            naverApi.getBlogs(title, DISPLAY_LIST_COUNT, startIndex).blogs
+        }
+
+    inner class BlogViewHolder(private val view: View) : RecyclerView.ViewHolder(view) {
```
ViewHolder 내부에 itemView를 가지고 있어서 view를 따로 선언해서 사용할 필요가 없습니다

gitpushyj replied:
반영했습니다!
40e4cdd5a04c66dec3b8e9e144ca511c9b103c21

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/blog/BlogListAdatper.kt:
```diff
@@ -0,0 +1,69 @@
+                fragmentActivity.startWebView(blog.link)
+            })
+        }
+        if (position > itemCount - PAGING_OFFSET) loadMore(itemCount)
```
이 방식은 좋지 않습니다.
아래로 스크롤을 빠르게 하면 같은 itemCount로 여러개의 api를 요청하게 됩니다.
다른 방식으로 구현해주세요

gitpushyj replied:
아하 그런 위험이 있었군요.
position이 가장 끝에 걸릴때만 추가 로딩하도록 로직을 수정했습니다.
말씀해주신 버그는 없어진것 같은데 더 좋은 방식이 있는지 모르겠네요..ㅠ

6c6794254332775a41f3f17c2e9806e74a6bc34b

sjjeong replied:
이 부분은 https://github.com/codepath/android_guides/wiki/Endless-Scrolling-with-AdapterViews-and-RecyclerView 내용 참고해보시면 좋습니다.

gitpushyj replied:
감사합니다. 한 번 정독해봐야겠네요 ㅎ

gitpushyj committed:
HW-1 응답 모델 리팩토링

gitpushyj committed:
HW-1 BaseSearchFragment 리팩토링

추상 프로퍼티가 아니라, 생성자로 layoutId를 주입받기 위함

gitpushyj committed:
HW-1 프로퍼티명 리팩토링

listView -> recyclerView

gitpushyj committed:
HW-1 onCreateView 기능 축소

onCreateView에서는 view 생성하는 역할만 담당하기 위함.
나머지 작업들은 onActivityCreated()에서.

gitpushyj committed:
HW-1 BaseSearchFragment 리팩토링

onActivityCreated에서 참조 가능한 fragmentView 프로퍼티 제거.
onActivityCreated에서 view로 접근 가능.

gitpushyj committed:
HW-1 불필요한 의존성 라이브러리 제거

gitpushyj committed:
HW-1 fetchSearchedList() 함수 리팩토링

gitpushyj committed:
HW-1 ViewHolder 리팩토링

RecyclerView.ViewHolder 내부에서 view가지고 있기 때문에 따로 선언해서 가지고 있을 필요 없음

gitpushyj committed:
HW-1 무한스크롤 로직 변경

sjjeong commented on wooooooak/StudyApp/app/src/main/java/wooooooak/com/studyapp/ui/blog/BlogFragment.kt:
```diff
@@ -0,0 +1,13 @@
+import wooooooak.com.studyapp.R
+import wooooooak.com.studyapp.ui.base.BaseSearchFragment
+
+class BlogFragment(layoutId: Int = R.layout.fragment_blog) : BaseSearchFragment(layoutId) {
```
```kotlin
class BlogFragment : BaseSearchFragment(R.layout.fragment_blog) {
```
이렇게 사용해도 됩니다.

gitpushyj replied:
넵 반영했습니다!

6e87af049b42b7fe21e3a9b39ab3609bc0a77ed8

gitpushyj committed:
HW-1 Fragment 리팩토링

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋

sjjeong commented:
@JayYi 확인후 Approve 해주세요
